src/analyzer/analyzerrhythm.cpp

```diff
@@ -0,0 +1,651 @@
+        for (auto &testHierarchy : pulseHierarchies) {
+            // handle standart case in which multiple of divisible by
+            // the last element already in the hierarchy
+            if (pulseMultiples[i] % testHierarchy.back() == 0) {
+                testHierarchy.push_back(pulseMultiples[i]);
+                pulseBelongs = true;
+            } else {
+                // handle edge cases in which multiple is not divisible
+                // by the last element but might be from some element before
+                auto last = testHierarchy.end() -1;
+                while (last != testHierarchy.begin()){
+                    if (pulseMultiples[i] % (*last) == 0) {
+                        pulseBelongs = true;
+                        std::vector<int> hierarchyCopy;
+                        std::copy(testHierarchy.begin(), last+1, std::back_inserter(hierarchyCopy));
+                        hierarchyCopy.push_back(pulseMultiples[i]);
+                        // this checks that two hierarchies have not reduced to the same
+                        // ex pulse = 20, with <2,4,12> and <2,4,8> both will reduce to <2,4>
+                        if( hierarchyCopy == pulseHierarchies.back()) {
+                            break;
+                        }
+                        pulseHierarchies.push_back(hierarchyCopy);
+                        break;
+                        
+                    }
+                    last -= 1;
+                }
+            }
+        }
+        // handle cases in which multiple does not belong in any hierarchy
+        if (!pulseBelongs) {
+            pulseHierarchies.push_back(lowestPulse);
+            pulseHierarchies.back().push_back(pulseMultiples[i]);
+        }
+    }
+    return pulseHierarchies;
+}
+
+int AnalyzerRhythm::computeNoveltyCurve() {
+    NoveltyCurveProcessor nc(static_cast<float>(m_iSampleRate), kNoveltyCurveWindow, kNoveltyCurveCompressionConstant);
+    m_noveltyCurve = nc.spectrogramToNoveltyCurve(m_spectrogram);
+    return m_spectrogram.size();
+}
+
+void AnalyzerRhythm::computeTempogramByDFT() {
+    auto hannWindow = std::vector<float>(m_tempogramWindowLength, 0.0);
+    WindowFunction::hanning(&hannWindow[0], m_tempogramWindowLength);
+    SpectrogramProcessor spectrogramProcessor(m_tempogramWindowLength,
+            m_tempogramFftLength, m_tempogramHopSize);
+    Spectrogram tempogramDFT = spectrogramProcessor.process(
+            &m_noveltyCurve[0], m_noveltyCurve.size(), &hannWindow[0]);
+    // convert y axis to bpm
+    int tempogramMinBin = (std::max(static_cast<int>(floor(((m_tempogramMinBPM/60.0)
+            /m_tempogramInputSampleRate)*m_tempogramFftLength)), 0));
+    int tempogramMaxBin = (std::min(static_cast<int>(ceil(((m_tempogramMaxBPM/60.0)
+            /m_tempogramInputSampleRate)*m_tempogramFftLength)), m_tempogramFftLength/2));
+    int binCount = tempogramMaxBin - tempogramMinBin + 1;
+    double highest;
+    double bestBpm;
+    int bin;
```

Comment:
Set but not used. Please always build with `-DDEBUG_ASSERTIONS_FATAL=ON -DWARNINGS_FATAL=ON` to catch these issues during development.